File: src/taskforge/client/tasks.py
from taskforge.client.base import ModelClient


class TaskClient(ModelClient):
    plural_name = "tasks"

    def next(self, context: str = ""):
        params = {"context": context} if context else None
        self.logger.debug("getting next Task in context: %s", context)
        return self.client.get("/api/v1/tasks/next", params=params)

    def search(self, query: str):
        params = {"q": query}
        self.logger.debug("searching for %s with query: %s", self.plural_name, query)
        return self.list(params=params)

    def complete(self, id: str):
        self.logger.debug("completing task: %s", id)
        return self.client.put(f"/api/v1/tasks/{id}/complete")

File: src/client/tasks.ts
import { ModelClient } from './model-client';

export interface Task {
    id: string; // UUID

    title: string;

    owner: string; // UUID

    source: string;
    priority: number;
    context: string;

    createdDate: string | Date;
    completedDate: string | Date | null;
}

export interface NewTask {
    title: string;

    priority?: number;
    context?: string;
}

export function isTask(data: any): data is Task {
    return (
        data &&
        data.title !== undefined &&
        data.owner !== undefined &&
        data.source !== undefined &&
        data.priority !== undefined &&
        data.context !== undefined &&
        data.createdDate !== undefined
    );
}

export class TaskClient extends ModelClient<Task, NewTask> {
    pluralName = 'tasks';
    validator = (data: any): Task => {
        if (!isTask(data)) {
            throw new Error(`Got unexpected Task data!`);
        }

        return data;
    };

    async search(query: string): Promise<Task[]> {
        return await this.list(0, { query });
    }

    async complete(id: string): Promise<null> {
        try {
            const url = `/${this.version}/${this.pluralName}/${id}`;
            await this.client.put(url);
            return null;
        } catch (e) {
            this.handleError(e);
            throw e;
        }
    }

    async next(context?: string): Promise<Task> {
        try {
            const url = context
                ? `/${this.version}/${this.pluralName}/next?context=${context}`
                : `/${this.version}/${this.pluralName}/next`;
            const response = await this.client.get(url);
            return this.validator(response.data);
        } catch (e) {
            this.handleError(e);
            throw e;
        }
    }
}
